### PR TITLE
refactor(explain): introduce `formatter_debug_plan_node` for convenient display of plan node properties

### PR DESCRIPTION
**Pull request by richardchien**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

We now have many bool properties for each stream plan node, including `append_only` and `eowc` for many nodes and `window`/`interval` for `StreamHashJoin`. Previously we display these properties by change the node name, e.g. `StreamAppendOnlyHashAgg`, but this approach cannot handle nodes with multiple properties. This PR changes the display behavior to `StreamHashAgg [append_only, eowc]`, and introduces a macro to ease the construction of the property list:

```rust
let mut builder = formatter_debug_plan_node!(
    f, "StreamHashJoin"
    , { "window", self.left().watermark_columns().contains(ljk) && self.right().watermark_columns().contains(rjk) }
    , { "interval", self.clean_left_state_conjunction_idx.is_some() && self.clean_right_state_conjunction_idx.is_some() }
    , { "append_only", self.is_append_only }
);
```

```rust
let mut builder = formatter_debug_plan_node!(
    f, "StreamHashAgg"
    , { "append_only", self.input().append_only() }
    , { "eowc", self.emit_on_window_close }
);
```

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- ~~[ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).~~
- ~~[ ] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)~~
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/frontend/planner_test/tests/testdata/output/watermark.yaml [hunk 1]
```diff
@@ -88,7 +88,7 @@
       └─LogicalScan { table: t2, columns: [t2.ts, t2.v1, t2.v2, t2._row_id] }
   stream_plan: |
     StreamMaterialize { columns: [t1_ts, t1_v1, t1_v2, t2_ts, t2_v1, t2_v2, t1._row_id(hidden), t2._row_id(hidden)], stream_key: [t1._row_id, t2._row_id, t1_v1], pk_columns: [t1._row_id, t2._row_id, t1_v1], pk_conflict: "NoCheck", watermark_columns: [t1_ts, t2_ts] }
-    └─StreamIntervalJoin { type: LeftOuter, predicate: t1.v1 = t2.v1 AND (t1.ts >= $expr2) AND ($expr1 <= t2.ts), conditions_to_clean_left_state_table: (t1.ts >= $expr2), conditions_to_clean_right_state_table: ($expr1 <= t2.ts), output_watermarks: [t1.ts, t2.ts], output: [t1.ts, t1.v1, t1.v2, t2.ts, t2.v1, t2.v2, t1._row_id, t2._row_id] }
+    └─StreamHashJoin [interval] { type: LeftOuter, predicate: t1.v1 = t2.v1 AND (t1.ts >= $expr2) AND ($expr1 <= t2.ts), conditions_to_clean_left_state_table: (t1.ts >= $expr2), conditions_to_clean_right_state_table: ($expr1 <= t2.ts), output_watermarks: [t1.ts, t2.ts], output: [t1.ts, t1.v1, t1.v2, t2.ts, t2.v1, t2.v2, t1._row_id, t2._row_id] }
       ├─StreamExchange { dist: HashShard(t1.v1) }
```
Comment (chenzl25) [inline src/frontend/planner_test/tests/testdata/output/watermark.yaml:92]: Ditto.

File: src/frontend/planner_test/tests/testdata/output/watermark.yaml [hunk 2]
```diff
@@ -71,7 +71,7 @@
     select t1.ts as t1_ts, t1.v1 as t1_v1, t1.v2 as t1_v2 from t1 where exists (select * from t2 where t1.ts = t2.ts);
   stream_plan: |
     StreamMaterialize { columns: [t1_ts, t1_v1, t1_v2, t1._row_id(hidden)], stream_key: [t1._row_id, t1_ts], pk_columns: [t1._row_id, t1_ts], pk_conflict: "NoCheck", watermark_columns: [t1_ts] }
-    └─StreamWindowJoin { type: LeftSemi, predicate: t1.ts = t2.ts, output_watermarks: [t1.ts], output: all }
+    └─StreamHashJoin [window] { type: LeftSemi, predicate: t1.ts = t2.ts, output_watermarks: [t1.ts], output: all }
```
[the remark this answers is not in the document]
Comment (richardchien) [reply src/frontend/planner_test/tests/testdata/output/watermark.yaml:74]: Where is the document, can't find it 🥵

[the remark this answers is not in the document]
Comment (chenzl25) [reply src/frontend/planner_test/tests/testdata/output/watermark.yaml:74]: It hasn't been released. Never mind.

**Review by chenzl25 — APPROVED** [review]
LGTM

File: src/frontend/src/optimizer/plan_node/utils.rs [hunk 3]
```diff
@@ -181,3 +181,26 @@ impl fmt::Debug for IndicesDisplay<'_> {
...
 }
+
+/// Call `debug_struct` on the given formatter to create a debug struct builder.
+/// If a property list is provided, properties in it will be added to the struct name according to
+/// the condition of that property.
+macro_rules! formatter_debug_plan_node {
+    ($formatter:ident, $name:literal) => {
+        $formatter.debug_struct($name)
+    };
+    ($formatter:ident, $name:literal, $( { $prop:literal, $cond:expr } ), *) => {
```
Comment (BugenZhao) [inline src/frontend/src/optimizer/plan_node/utils.rs:192]:
```suggestion
    ($formatter:ident, $name:literal $(, { $prop:literal, $cond:expr } )* $(,)?) => {
```

Can we unify them in one branch?

**Review by BugenZhao — APPROVED** [review]

**Pull request merged**